src/cljam/algo/pileup.clj

```diff
@@ -35,34 +35,34 @@
   [^long start ^long end alns]
   (seq-step start end 0 alns))
 
-(defn- quals-at-ref
-  [idx ^String qual]
-  (let [empty-qual? (and (= (.length qual) 1)
-                         (= (.charAt qual 0) \*))]
-    (if empty-qual?
-      (vec (repeat (count idx) 93)) ;; \~
-      (mapv (fn [[_ x]]
-              (if (number? x)
-                (qual/fastq-char->phred-byte (.charAt qual x))
-                93))
-            idx))))
-
-(defn- seqs-at-ref
-  [idx ^String s]
-  (mapv (fn [[op x xs]]
-          (let [c (if (number? x) (.charAt s x) x)]
-            (case op
-              :m [c]
-              :d [c xs]
-              :i [c (subs s (first xs) (last xs))]))) idx))
-
 (defn index-cigar
   "Align bases and base quality scores with the reference coordinate."
   [^SAMAlignment aln]
-  (let [idx (cigar/to-index (.cigar aln))]
-    (assoc aln
-           :seqs-at-ref (seqs-at-ref idx (:seq aln))
-           :quals-at-ref (quals-at-ref idx (.qual aln)))))
+  (let [idx (cigar/to-index (.cigar aln))
+        ^String seqs (:seq aln)
+        ^String quals (.qual aln)
+        empty-qual? (and (= (.length quals) 1)
+                         (= (.charAt quals 0) \*))]
+    (loop [i idx
+           seqs-at-ref (transient [])
+           quals-at-ref (transient [])]
+      (let [[op x xs] (first i)]
+        (if op
```

Comment:
Destructuring makes it possible to describe first and rest parts simultaneously.

```clojure
    (loop [[[op x xs] & idx] idx
           seqs-at-ref (transient [])
           quals-at-ref (transient [])]
      (if op
```

Reply:
Thanks! Fixed!